kvdb/postgres/readwrite_bucket.go

```diff
@@ -0,0 +1,354 @@
+package postgres
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/jackc/pgx/v4"
+
+	"github.com/btcsuite/btcwallet/walletdb"
+)
+
+// readWriteBucket stores the bucket id and the buckets transaction.
+type readWriteBucket struct {
+	// id is used to identify the bucket. If id is null, it refers to the
+	// root bucket.
+	id *int64
+
+	// tx holds the parent transaction.
+	tx *readWriteTx
+}
+
+// newReadWriteBucket creates a new rw bucket with the passed transaction
+// and bucket id.
+func newReadWriteBucket(tx *readWriteTx, id *int64) *readWriteBucket {
+	return &readWriteBucket{
+		id: id,
+		tx: tx,
+	}
+}
+
+// NestedReadBucket retrieves a nested read bucket with the given key.
+// Returns nil if the bucket does not exist.
+func (b *readWriteBucket) NestedReadBucket(key []byte) walletdb.ReadBucket {
+	return b.NestedReadWriteBucket(key)
+}
+
+func parentSelector(id *int64) string {
+	if id == nil {
+		return "parent_id IS NULL"
+	}
+	return fmt.Sprintf("parent_id=%v", *id)
+}
+
+// ForEach invokes the passed function with every key/value pair in
+// the bucket. This includes nested buckets, in which case the value
+// is nil, but it does not include the key/value pairs within those
+// nested buckets.
+func (b *readWriteBucket) ForEach(cb func(k, v []byte) error) error {
+	rows, err := b.tx.tx.Query(
+		context.TODO(),
+		"SELECT key,value FROM kv WHERE "+parentSelector(b.id)+" ORDER BY key",
```

Comment:
I see what you mean, only I don't know if an abstraction layer actually makes it better. Especially because of these subtle differences, you get a set of arguments to flag those and it may not be that clear at the call site what is actually happening. There are other, more ORM style sql libraries such as https://github.com/go-pg/pg that allow queries to be written as a series of go calls, but I decided against that because it seemed heavier and perhaps not very necessary given that our db code is very localized.

A concrete example of the type of abstraction that you envision would be helpful.

Reply:
I gave this a quick attempt but also came to the conclusion that adding another abstraction layer over the SQL statements probably doesn't add that much clarity at this point...

Reply:
Ok, closing the thread for now then.